## 📋 Quality Gate Results

| Check | Status |
|-------|--------|
| Knip | ✅ success |
| Lint | ❌ failure |
| Slop | ✅ success |
| Build | ✅ success |
| Infra Tests | ❌ skipped |
| Unit Tests | ❌ skipped |
| Component Tests | ❌ skipped |
| Perf Tests | ❌ skipped |
| Visual Tests | ❌ skipped |


### ❌ Lint Failure Details
```text

> hrm@0.30.0 lint /home/runner/work/hrm/hrm
> eslint app/ components/ constants/ context/ hooks/ lib/ services/ tests/ types/ utils/ server.ts proxy.ts --cache


/home/runner/work/hrm/hrm/app/client/experimental/components/ZoneDistribution.tsx
  20:10  error  Replace `⏎··(acc,·config)·=>·({·...acc,·[config.label]:·config.color·}),⏎·` with `(acc,·config)·=>·({·...acc,·[config.label]:·config.color·}),`  prettier/prettier
  23:1   error  Replace `····` with `··`                                                                                                                         prettier/prettier
  24:1   error  Delete `··`                                                                                                                                      prettier/prettier
  25:1   error  Replace `··}·as·Record<string,·string>⏎` with `}·as·Record<string,·string>`                                                                      prettier/prettier
  44:12  error  Insert `(⏎······`                                                                                                                                prettier/prettier
  45:7   error  Insert `··`                                                                                                                                      prettier/prettier
  46:1   error  Insert `··`                                                                                                                                      prettier/prettier
  47:7   error  Insert `··`                                                                                                                                      prettier/prettier
  48:1   error  Replace `······` with `········`                                                                                                                 prettier/prettier
  49:9   error  Replace `const·percentage·=` with `··const·percentage·=⏎···········`                                                                             prettier/prettier
  50:9   error  Insert `··`                                                                                                                                      prettier/prettier
  51:1   error  Insert `··`                                                                                                                                      prettier/prettier
  52:1   error  Replace `··········` with `············`                                                                                                         prettier/prettier
  53:1   error  Insert `··`                                                                                                                                      prettier/prettier
  54:1   error  Insert `··`                                                                                                                                      prettier/prettier
  55:11  error  Insert `··`                                                                                                                                      prettier/prettier
  56:9   error  Insert `··`                                                                                                                                      prettier/prettier
  57:1   error  Replace `······` with `········`                                                                                                                 prettier/prettier
  58:1   error  Insert `··`                                                                                                                                      prettier/prettier
  59:1   error  Insert `··`                                                                                                                                      prettier/prettier
  60:7   error  Insert `··`                                                                                                                                      prettier/prettier
  61:7   error  Insert `··`                                                                                                                                      prettier/prettier
  62:1   error  Replace `········` with `··········`                                                                                                             prettier/prettier
  63:11  error  Insert `··`                                                                                                                                      prettier/prettier
  64:1   error  Insert `··`                                                                                                                                      prettier/prettier
  65:9   error  Insert `··`                                                                                                                                      prettier/prettier
  66:7   error  Replace `}` with `··})⏎····`                                                                                                                     prettier/prettier

✖ 27 problems (27 errors, 0 warnings)
  27 errors and 0 warnings potentially fixable with the `--fix` option.

 ELIFECYCLE  Command failed with exit code 1.
```
### ❌ Infrastructure Test Failure Details
```text
Log file not found.
```
### ❌ Unit Test Failure Details
```text
Log file not found.
```
### ❌ Component Test Failure Details
```text
Log file not found.
```
### ❌ Visual Test Failure Details
```text
Log file not found.
```
> 💡 **Tip:** Download the `visual-test-logs-8335` artifact to view the full interactive Playwright report.

### ❌ Performance Test Failure Details
```text
Log file not found.
```
⚠️ **Some checks failed.** Full logs available in [workflow artifacts](https://github.com/arii/hrm/actions/runs/22013093204).

---
> Report generated for commit: `eb94f6bb0b70a75307a7740b4e6c5c269bb3810d`

--- a/app/client/experimental/components/ZoneDistribution.tsx
+++ b/app/client/experimental/components/ZoneDistribution.tsx
@@ -4,7 +4,11 @@
 import React, { useMemo } from 'react'
 import { Card, CardContent, Typography, Box, useTheme } from '@mui/material'
 import { PieChart, Pie, Cell, ResponsiveContainer, Tooltip } from 'recharts'
-import { HrZoneName, HR_ZONE_VISUAL_CONFIG } from '@/lib/shared/hr-zones'
+import {
+  HrZoneName,
+  HR_ZONE_VISUAL_CONFIG,
+  HR_ZONE_ORDER,
+} from '@/lib/shared/hr-zones'
 import { formatDuration } from '@/lib/utils'
 
 // --- Types ---
@@ -59,12 +63,10 @@ const ZoneDistribution: React.FC<ZoneDistributionProps> = ({
         // Filter out zero values to keep the chart clean
         .filter((item) => item.value > 0)
         // Sort by intensity (Max to Idle/Recovery)
-        .sort((a, b) => {
-          const order = Object.values(HR_ZONE_VISUAL_CONFIG)
-            .map((c) => c.label as string)
-            .reverse()
-          return order.indexOf(a.name) - order.indexOf(b.name)
-        })
+        .sort(
+          (a, b) =>
+            HR_ZONE_ORDER.indexOf(a.name) - HR_ZONE_ORDER.indexOf(b.name)
+        )
     )
   }, [timeInZones, totalDuration, theme.palette.grey])
 
@@ -90,6 +92,40 @@ const ZoneDistribution: React.FC<ZoneDistributionProps> = ({
 
   return (
     <Card elevation={3}>
+      {/* Visually hidden table for screen reader accessibility */}
+      <Box
+        component="table"
+        sx={{
+          position: 'absolute',
+          width: '1px',
+          height: '1px',
+          padding: 0,
+          margin: '-1px',
+          overflow: 'hidden',
+          clip: 'rect(0, 0, 0, 0)',
+          whiteSpace: 'nowrap',
+          border: 0,
+        }}
+      >
+        <caption>Heart Rate Zone Distribution Data Table</caption>
+        <thead>
+          <tr>
+            <th scope="col">Zone</th>
+            <th scope="col">Duration</th>
+            <th scope="col">Percentage</th>
+          </tr>
+        </thead>
+        <tbody>
+          {data.map((item) => (
+            <tr key={`sr-row-${item.name}`}>
+              <td>{item.name}</td>
+              <td>{item.formattedTime}</td>
+              <td>{item.percentage}%</td>
+            </tr>
+          ))}
+        </tbody>
+      </Box>
+
       <CardContent>
         <Typography variant="h6" component="h2" gutterBottom fontWeight="bold">
           Heart Rate Zone Distribution

--- a/app/client/experimental/components/ZoneDistribution.tsx
+++ b/app/client/experimental/components/ZoneDistribution.tsx
@@ -17,13 +17,10 @@ interface ZoneDistributionProps {
 // Map HR Zone labels to colors using canonical config
 const ZONE_COLOR_MAP: Record<string, string> = Object.values(
   HR_ZONE_VISUAL_CONFIG
-).reduce(
-  (acc, config) => ({ ...acc, [config.label]: config.color }),
-  {
-    [HrZoneName.NoData]: '#e0e0e0',
-    [HrZoneName.Unknown]: '#9e9e9e',
-  } as Record<string, string>
-)
+).reduce((acc, config) => ({ ...acc, [config.label]: config.color }), {
+  [HrZoneName.NoData]: '#e0e0e0',
+  [HrZoneName.Unknown]: '#9e9e9e',
+} as Record<string, string>)
 
 // Strict time formatter: "MM:SS" (e.g., "25:40")
 const formatTime = (seconds: number): string => {
@@ -41,29 +38,32 @@ const ZoneDistribution: React.FC<ZoneDistributionProps> = ({
 
   // Transform data for Recharts and list display
   const data = useMemo(() => {
-    return Object.entries(timeInZones)
-      .filter(
-        ([zone]) => zone !== HrZoneName.NoData && zone !== HrZoneName.Unknown
-      )
-      .map(([zone, time]) => {
-        const percentage = totalDuration > 0 ? (time / totalDuration) * 100 : 0
-        return {
-          name: zone,
-          value: time,
-          percentage: parseFloat(percentage.toFixed(1)),
-          formattedTime: formatTime(time),
-          color: ZONE_COLOR_MAP[zone] || theme.palette.grey[500],
-        }
-      })
-      // Filter out zero values to keep the chart clean
-      .filter((item) => item.value > 0)
-      // Sort by intensity (Max to Idle/Recovery)
-      .sort((a, b) => {
-        const order = Object.values(HR_ZONE_VISUAL_CONFIG)
-          .map((c) => c.label as string)
-          .reverse()
-        return order.indexOf(a.name) - order.indexOf(b.name)
-      })
+    return (
+      Object.entries(timeInZones)
+        .filter(
+          ([zone]) => zone !== HrZoneName.NoData && zone !== HrZoneName.Unknown
+        )
+        .map(([zone, time]) => {
+          const percentage =
+            totalDuration > 0 ? (time / totalDuration) * 100 : 0
+          return {
+            name: zone,
+            value: time,
+            percentage: parseFloat(percentage.toFixed(1)),
+            formattedTime: formatTime(time),
+            color: ZONE_COLOR_MAP[zone] || theme.palette.grey[500],
+          }
+        })
+        // Filter out zero values to keep the chart clean
+        .filter((item) => item.value > 0)
+        // Sort by intensity (Max to Idle/Recovery)
+        .sort((a, b) => {
+          const order = Object.values(HR_ZONE_VISUAL_CONFIG)
+            .map((c) => c.label as string)
+            .reverse()
+          return order.indexOf(a.name) - order.indexOf(b.name)
+        })
+    )
   }, [timeInZones, totalDuration, theme.palette.grey])
 
   if (data.length === 0) {

--- a/app/client/experimental/components/ZoneDistribution.tsx
+++ b/app/client/experimental/components/ZoneDistribution.tsx
@@ -5,6 +5,7 @@ import React, { useMemo } from 'react'
 import { Card, CardContent, Typography, Box, useTheme } from '@mui/material'
 import { PieChart, Pie, Cell, ResponsiveContainer, Tooltip } from 'recharts'
 import { HrZoneName, HR_ZONE_VISUAL_CONFIG } from '@/lib/shared/hr-zones'
+import { formatDuration } from '@/lib/utils'
 
 // --- Types ---
 interface ZoneDistributionProps {
@@ -24,10 +25,11 @@ const ZONE_COLOR_MAP: Record<string, string> = Object.values(
 
 // Strict time formatter: "MM:SS" (e.g., "25:40")
 const formatTime = (seconds: number): string => {
-  const safeSeconds = Math.max(0, Math.floor(seconds))
-  const mins = Math.floor(safeSeconds / 60)
-  const secs = safeSeconds % 60
-  return `${mins}:${secs.toString().padStart(2, '0')}`
+  return formatDuration(seconds, {
+    unit: 'seconds',
+    format: 'MM:SS',
+    noPadMinutes: true,
+  })
 }
 
 const ZoneDistribution: React.FC<ZoneDistributionProps> = ({
@@ -104,6 +106,10 @@ const ZoneDistribution: React.FC<ZoneDistributionProps> = ({
             width={{ xs: '100%', sm: '50%' }}
             height={200}
             position="relative"
+            role="img"
+            aria-label={`Donut chart showing heart rate zone distribution. Total duration: ${formatTime(
+              totalDuration
+            )}.`}
           >
             <ResponsiveContainer width="100%" height="100%">
               <PieChart>

--- a/lib/shared/hr-zones.ts
+++ b/lib/shared/hr-zones.ts
@@ -106,6 +106,13 @@ export const HR_ZONE_VISUAL_CONFIG = {
 } as const
 
 /**
+ * Pre-calculated order of zone labels for sorting purposes (highest intensity first).
+ */
+export const HR_ZONE_ORDER = Object.values(HR_ZONE_VISUAL_CONFIG)
+  .map((config) => config.label as string)
+  .reverse()
+
+/**
  * Calculates the percentage of max HR and the corresponding zone (0-5) based on Max HR.
  * @param currentHr - Current heart rate in BPM.
  * @param maxHr - Max Heart Rate.

--- a/lib/utils.ts
+++ b/lib/utils.ts
@@ -35,9 +35,10 @@ export const formatDuration = (
   options: {
     unit: 'seconds' | 'milliseconds'
     format?: 'HH:MM:SS' | 'MM:SS'
+    noPadMinutes?: boolean
   }
 ): string => {
-  const { unit, format = 'HH:MM:SS' } = options
+  const { unit, format = 'HH:MM:SS', noPadMinutes = false } = options
 
   if (isNaN(duration) || duration < 0) {
     if (format === 'HH:MM:SS') {
@@ -50,9 +51,10 @@ export const formatDuration = (
     unit === 'milliseconds' ? Math.floor(duration / 1000) : duration
 
   if (format === 'MM:SS') {
-    const minutes = Math.floor(totalSeconds / 60)
-      .toString()
-      .padStart(2, '0')
+    const minsRaw = Math.floor(totalSeconds / 60)
+    const minutes = noPadMinutes
+      ? minsRaw.toString()
+      : minsRaw.toString().padStart(2, '0')
     const seconds = Math.floor(totalSeconds % 60)
       .toString()
       .padStart(2, '0')

--- a/tests/unit/app/client/experimental/components/ZoneDistribution.test.tsx
+++ b/tests/unit/app/client/experimental/components/ZoneDistribution.test.tsx
@@ -60,15 +60,22 @@ describe('ZoneDistribution', () => {
       [HrZoneName.Cardio]: 10,
     }
     render(<ZoneDistribution timeInZones={timeInZones} totalDuration={100} />)
-    expect(screen.getByText(HrZoneName.WarmUp)).toBeInTheDocument()
-    expect(screen.getByText('1:00')).toBeInTheDocument() // 60s formatted as MM:SS
-    expect(screen.getByText(/60%/)).toBeInTheDocument()
-    expect(screen.getByText(HrZoneName.FatBurn)).toBeInTheDocument()
-    expect(screen.getByText('0:30')).toBeInTheDocument() // 30s formatted as MM:SS
-    expect(screen.getByText(/30%/)).toBeInTheDocument()
-    expect(screen.getByText(HrZoneName.Cardio)).toBeInTheDocument()
-    expect(screen.getByText('0:10')).toBeInTheDocument() // 10s formatted as MM:SS
-    expect(screen.getByText(/10%/)).toBeInTheDocument()
+    // Multiple instances due to accessible table and visible legend
+    expect(
+      screen.getAllByText(HrZoneName.WarmUp).length
+    ).toBeGreaterThanOrEqual(1)
+    expect(screen.getAllByText('1:00').length).toBeGreaterThanOrEqual(1)
+    expect(screen.getAllByText(/60%/).length).toBeGreaterThanOrEqual(1)
+    expect(
+      screen.getAllByText(HrZoneName.FatBurn).length
+    ).toBeGreaterThanOrEqual(1)
+    expect(screen.getAllByText('0:30').length).toBeGreaterThanOrEqual(1)
+    expect(screen.getAllByText(/30%/).length).toBeGreaterThanOrEqual(1)
+    expect(
+      screen.getAllByText(HrZoneName.Cardio).length
+    ).toBeGreaterThanOrEqual(1)
+    expect(screen.getAllByText('0:10').length).toBeGreaterThanOrEqual(1)
+    expect(screen.getAllByText(/10%/).length).toBeGreaterThanOrEqual(1)
   })
 
   it('does not display zones with no time', () => {
@@ -77,7 +84,9 @@ describe('ZoneDistribution', () => {
       [HrZoneName.WarmUp]: 100,
     }
     render(<ZoneDistribution timeInZones={timeInZones} totalDuration={100} />)
-    expect(screen.getByText(HrZoneName.WarmUp)).toBeInTheDocument()
+    expect(
+      screen.getAllByText(HrZoneName.WarmUp).length
+    ).toBeGreaterThanOrEqual(1)
     expect(screen.queryByText(HrZoneName.FatBurn)).not.toBeInTheDocument()
   })
 
@@ -103,7 +112,7 @@ describe('ZoneDistribution', () => {
     }
     render(<ZoneDistribution timeInZones={timeInZones} totalDuration={120} />)
     expect(screen.getByText('Total')).toBeInTheDocument()
-    // 2:00 appears twice (center and legend). We verify at least one exists.
+    // Multiple instances due to center label, legend, and table
     const timeElements = screen.getAllByText('2:00')
     expect(timeElements.length).toBeGreaterThanOrEqual(1)
   })

